solutions/0400-0499/0403-Frog-Jump.md

```diff
@@ -0,0 +1,187 @@
+        map[0].insert(1);
+        
+        // iterating every stone positions through $stones$ array
+        for (int i = 0; i < n; i++) {
+            int currstone = stones[i];
+            
+            // iterating each moves
+            set<int> st = map[currstone];
+            
+            for (auto x : st) {
+                // pos is stone's positon after applying move x
+                int pos = currstone + x;
+                
+                // if position of stone is equal to final position then frog will win 
+                if (pos == stones[stones.size() - 1]) {
+                     return true;
+                }
+  
+                // first we will check pos is available in our stones array
+                if (map.find(pos) != map.end()) {
+                    // if the frog's last jump was x units, its next jump must be either x - 1, x, or x + 1 units
+                    if (x - 1 > 0) {  
+                        map[pos].insert(x - 1);  
+                    }
+                    
+                    map[pos].insert(x);
+                    map[pos].insert(x + 1); 
+                }
+            }
+        }
+      
+        return false;
+    }
+};
+  
+```
+</TabItem>
+</Tabs>
+
+## Approach 2: Dynamic Programming  
+  We apply Recursion + Memoization to find out the solution. There are three possible recursion call for $k-1$, $k$, $k+1$ moves.  
+  There are some base condtions, we have to facus on 
+- if move k is less then zero, then we can not make a jump at that position.
+- Now, we have to find that whatever $move$ $(stones[idx-1] + k)$ that has been made is present in our $stones$ array or not
+- Also we are answer in our dp and if $dp[idx][k]!=-1$, it means our answer is already present in dp
+- Now, final base condition will check if we reach last position or not  
+  
+Time Complexity is $O(n*n)$, where $n$ is the length of the $stones$ array.  
+Space Complexity is $O(n*n)$
+
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@Kavita613"/>
+
+```cpp  
+class Solution {  
+public:
+    // Create Globally 2d dp array 
+    int dp[2002][2002];
+    
+    bool solve(int idx, int k, vector<int>& stones, unordered_map<int, int>& m)
+    {   
+        // k is number of moves 
+        if (k <= 0) return false;
+        
+        // checking k + stones[idx - 1] is available position in our stones array or not
+        if (m.find(k + stones[idx - 1]) == m.end()) {
+            return false;  
+        } else {
+            idx = m[k + stones[idx - 1]];
+        }
+        if (dp[idx][k] != -1) {
+            return dp[idx][k];
+        }
+                  
+        // when we reach at end of the positions we will return true
+        if (idx == stones.size() - 1) {
+            return true;
+        }
+        // calling recursive function for k-1, k, k+1 moves
+        dp[idx][k]  = solve(idx + 1, k-1, stones, mp) ||  solve(idx + 1, k, stones, mp) ||  solve(idx + 1, k+1, stones, mp);
```

Comment:
This line would fail. Please change to 

```cpp
dp[idx][k]  = solve(idx + 1, k - 1, stones, m) ||  solve(idx + 1, k, stones, m) ||  solve(idx + 1, k + 1, stones, m);
```